_posts/2019-11-13-variational-convnet/variational_convnet.Rmd

```diff
@@ -0,0 +1,619 @@
+$ observationId <int> 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 13, 14, 17, 18, 19, 2…
+```
+
+```{r}
+testData %>% glimpse()
+```
+
+```
+Observations: 69
+Variables: 6
+$ experiment    <int> 11, 12, 15, 16, 32, 33, 42, 43, 52, 53, 56, 57, 11, …
+$ userId        <int> 6, 6, 8, 8, 16, 16, 21, 21, 26, 26, 28, 28, 6, 6, 8,…
+$ activity      <int> 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 8, 8…
+$ data          <list> [<data.frame[185 x 6]>, <data.frame[151 x 6]>, <dat…
+$ activityName  <fct> STAND_TO_SIT, STAND_TO_SIT, STAND_TO_SIT, STAND_TO_S…
+$ observationId <int> 11, 12, 15, 16, 31, 32, 41, 42, 51, 52, 55, 56, 71, …
+```
+
+
+The code required to arrive at this stage (copied from Nick's post) may be found in the appendix at the bottom of this page.
+
+## Training pipeline
+
+The dataset in question is small enough to fit in memory -- but yours might not be, so it can't hurt to see some streaming in action. Besides, it's probably safe to say that with TensorFlow 2.0, [tfdatasets](https://tensorflow.rstudio.com/tools/tfdatasets/articles/introduction.html) pipelines are _the_ way to feed data to a model.
+
+Once the code listed in the appendix has run, the sensor data is to be found in `trainData$data`, a list column containing `data.frame`s where each row corresponds to a point in time and each column holds one of the measurements. However, not all time series (recordings) are of the same length; we thus follow the original post to pad all series to length `pad_size` (= 338). The expected shape of training batches will then be `(batch_size, pad_size, 6)`.
+
+We initially create our training dataset:
+
+```{r}
+train_x <- train_data$data %>% 
+  map(as.matrix) %>%
+  pad_sequences(maxlen = pad_size, dtype = "float32") %>%
+  tensor_slices_dataset() 
+
+train_y <- train_data$activity %>% 
+  one_hot_classes() %>% 
+  tensor_slices_dataset()
+
+train_dataset <- zip_datasets(train_x, train_y)
+train_dataset
+```
+
+```
+<ZipDataset shapes: ((338, 6), (6,)), types: (tf.float64, tf.float64)>
+```
+
+Then shuffle and batch it:
+
+
+```{r}
+n_train <- nrow(train_data)
+# the highest possible batch size for this dataset
+# chosen because it yielded the best performance
+# alternatively, experiment with e.g. different learning rates, ...
+batch_size <- n_train
+
+train_dataset <- train_dataset %>% 
+  dataset_shuffle(n_train) %>%
+  dataset_batch(batch_size)
+train_dataset
+```
+
+```
+<BatchDataset shapes: ((None, 338, 6), (None, 6)), types: (tf.float64, tf.float64)>
+```
+
+
+Same for the test data.
+
+```{r}
+test_x <- test_data$data %>% 
+  map(as.matrix) %>%
+  pad_sequences(maxlen = pad_size, dtype = "float32") %>%
+  tensor_slices_dataset() 
+
+test_y <- test_data$activity %>% 
+  one_hot_classes() %>% 
+  tensor_slices_dataset()
+
+n_test <- nrow(test_data)
+test_dataset <- zip_datasets(test_x, test_y) %>%
+  dataset_batch(n_test)
+
+```
+
+Using `tfdatasets` does not mean we cannot run a quick sanity check on our data:
```

Comment:
The double-negative above sounds a little off. Perhaps something ~ "We'll run a quick sanity check on our data (even though it's from `tfdatasets`)." or something of the like?